File: ios/UniversalTooltipModule.swift
import ExpoModulesCore

public class UniversalTooltipModule: Module {

  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('UniversalTooltip')` in JavaScript.
    Name("UniversalTooltip")
    
    View(UniversalTooltipView.self) {
      Events("onTap")
      Events("onDismiss")
      
      Prop("open") { (view, opened: Bool?) in
        view.opened = opened ?? false
      }
      Prop("side") { (view, side: ContentSide?) in
        view.side = side ?? .any
      }
      Prop("maxWidth") { (view, maxWidth: Double?) in
        view.maxWidth = maxWidth
      }
      Prop("arrowWidth") { (view, arrowWidth: Double) in
        view.arrowWidth = arrowWidth
      }
      Prop("arrowHeight") { (view, arrowWidth: Double) in
        view.arrowHeight = arrowWidth
      }
      Prop("text") { (view, text: String?) in
        view.text = text
      }
      Prop("presetAnimation") { (view, presetAnimation: PresetAnimation?) in
        view.presetAnimation = presetAnimation ?? .fadeIn
      }
      Prop("showDuration") { (view, showDuration: Double? ) in
        view.showDuration = (showDuration ?? 300)/1000
      }
      Prop("dismissDuration") { (view, dismissDuration: Double?) in
        view.dismissDuration = (dismissDuration ?? 300)/1000
      }
      Prop("containerStyle") { (view, containerStyle: ContainerStyle?) in
        view.containerStyle = containerStyle
      }
      Prop("textStyle") { (view, textStyle: TextStyle?) in
        view.textStyle = textStyle ?? TextStyle(fontSize: 13)
      }
      Prop("sideOffset") { (view, sideOffset: Double?) in
        view.sideOffset = sideOffset ?? Double(1)
      }
      Prop("disableTapToDismiss") { (view, disableTapToDismiss: Bool?) in
        view.disableTapToDismiss = disableTapToDismiss ?? false
      }
      Prop("disableDismissWhenTouchOutside") { (view, disableDismissWhenTouchOutside: Bool?) in
        view.disableDismissWhenTouchOutside = disableDismissWhenTouchOutside ?? false
      }
      Prop("borderRadius") { (view, borderRadius: Double?) in
        view.cornerRadius = borderRadius ?? CGFloat(0)
      }
      Prop("backgroundColor") { (view, backgroundColor: UIColor?) in
        view.bubbleBackgroundColor = backgroundColor ?? .clear
      }
    }
  }
}

File: android/src/main/java/expo/modules/universaltooltip/UniversalTooltipModule.kt
package expo.modules.universaltooltip

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.universaltooltip.records.ContainerStyle
import expo.modules.universaltooltip.enums.ContentSide
import expo.modules.universaltooltip.enums.PresetAnimation
import expo.modules.universaltooltip.records.TextStyle

class UniversalTooltipModule : Module() {
    // Each module class must implement the definition function. The definition consists of components
    // that describes the module's functionality and behavior.
    // See https://docs.expo.dev/modules/module-api for more details about available components.
    override fun definition() = ModuleDefinition {
        // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
        // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
        // The module will be accessible from `requireNativeModule('UniversalTooltip')` in JavaScript.
        Name("UniversalTooltip")

        // Enables the module to be used as a native view. Definition components that are accepted as part of
        // the view definition: Prop, Events.
        View(UniversalTooltipView::class) {
            Events(
                "onTap",
                "onDismiss",
            )
            Prop("open") { view: UniversalTooltipView, open: Boolean ->
                view.opened = open
            }
            Prop("side") { view: UniversalTooltipView, side: ContentSide ->
                view.side = side
            }
            Prop("text") { view: UniversalTooltipView, text: String ->
                view.text = text
            }
            Prop("maxWidth") { view: UniversalTooltipView, maxWidth: Int ->
                view.maxWidth = maxWidth
            }
            Prop("arrowWidth") { view: UniversalTooltipView, arrowWidth: Int ->
                view.arrowWidth = arrowWidth
            }
            Prop("arrowHeight") { view: UniversalTooltipView, arrowHeight: Int ->
                view.arrowHeight = arrowHeight
            }
            Prop("presetAnimation") { view: UniversalTooltipView, presetAnimation: PresetAnimation ->
                view.presetAnimation = presetAnimation
            }
            Prop("showDuration") { view: UniversalTooltipView, showDuration: Double ->
                view.showDuration = showDuration
            }
            Prop("containerStyle") { view: UniversalTooltipView, containerStyle: ContainerStyle ->
                view.containerStyle = containerStyle
            }
            Prop("textStyle") { view: UniversalTooltipView, textStyle: TextStyle? ->
                view.textStyle = textStyle
            }
            Prop("sideOffset") { view: UniversalTooltipView, sideOffset: Int ->
                view.sideOffset = sideOffset
            }
            Prop("disableTapToDismiss") { view: UniversalTooltipView, disableTapToDismiss: Boolean ->
                view.disableTapToDismiss = disableTapToDismiss
            }
            Prop("borderRadius") { view: UniversalTooltipView, borderRadius: Float ->
                view.borderRadius = borderRadius
            }
            Prop("backgroundColor") { view: UniversalTooltipView, backgroundColor: Int ->
                view.bgColor = backgroundColor
            }
            Prop("disableDismissWhenTouchOutside") { view: UniversalTooltipView, disableDismissWhenTouchOutside: Boolean ->
                view.disableDismissWhenTouchOutside = disableDismissWhenTouchOutside
            }

        }
    }
}
